### ISSUE TITLE
Risk Controls

### ISSUE THREAD

**Issue by anilhelvaci** [anilhelvaci/dapp-pool-lending-protocol#45]
Behaviors described in [Package-2](https://github.com/anilhelvaci/dapp-pool-lending-protocol/wiki/Work-Plan-For-Expand-Specification#package-2) should be implemented.

Risk Controls;
* Individual assets can be marked borrowable or non-borrowable
* Individual assets can be marked usable as collateral or not
* Limit can be set on supply accepted per collateral

**Comment by anilhelvaci** [comment]
## Think
I our meeting with Agoric team we agreed that the risk controls should not be a subject to governance voting. This leads me to think there two main paths we can take;
1. These risk controls are mutable and the power to mutate those variables belong to the creator. Or;
2. Risk controls are immutable and they're declared when adding the new pool to the protocol via governance voting.

I'm leaning towards to option two because giving such a power to the creator feels like an anti-pattern to me. What do you think @rowgraus ?

**Comment by rowgraus** [comment]
I agree with your POV around giving too much control to the contract creator being an anti-pattern.  However, for a production system we'd likely want to expose risk controls to *some* kind of governance body.  We avoided doing that for now to keep scope manageable.

Presumably, leaving the control with the contract creator for now would make it easier to upgrade later to have that control owned by a different governor?  If so, I'd argue for (1) vs. (2)

**Comment by anilhelvaci** [comment]
> Presumably, leaving the control with the contract creator for now would make it easier to upgrade later to have that control owned by a different governor? If so, I'd argue for (1) vs. (2)

In coding wise I don't think it would make it any easier the process of handing over the control to actual governance from the creator since governance mechanism updates parameters via a `paramManager` and contracts do not have the authority to update the parameters by themselves. But I think making the risk controls mutable will enable us to explore challenges we might face later when we integrate governance to these parameters. What I mean by challenges is something like, we start a pool where its protocol token is marked `usable as collateral` and after some time we mark it as `non-usable as collateral` and later we mark it as `usable as collateral` again. This may result in some challenges keeping the correct pool state. So I'll go for 1.

**Comment by anilhelvaci** [comment]
> Limit can be set on supply accepted per collateral

Let's say we have three pools in our protocol. Namely;
* VAN Pool, underlying asset `VAN` protocol token `AgVAN`
* PAN Pool, underlying asset `PAN` protocol token `AgPAN`
* HOP Pool, underlying asset `HOP` protocol token `AgHOP`

For the VAN Pool,
* `AgPAN` limit = 100 Units
* `AgHOP` limit = 150 Units

For the PAN Pool,
* `AgVAN` limit = 130 Units
* `AgHOP` limit = 90 Units

For the HOP Pool,
* `AgVAN` limit = 200 Units
* `AgPAN` limit = 50 Units

Or,

* Total limit for `AgVAN` used as collateral = 100 Units (PAN Pool + HOP Pool)
* Total limit for `AgPAN` used as collateral = 150 Units (VAN Pool + HOP Pool)
* Total limit for `AgHOP` used as collateral = 200 Units (VAN Pool + PAN Pool)

Which of the above approaches should we go for? @rowgraus

**Comment by anilhelvaci** [comment]
> * Total limit for AgVAN used as collateral = 100 Units (PAN Pool + HOP Pool)
> * Total limit for AgPAN used as collateral = 150 Units (VAN Pool + HOP Pool)
> * Total limit for AgHOP used as collateral = 200 Units (VAN Pool + PAN Pool)

After a brief discussion with @rowgraus, we've decided go with the above approach.

**Comment by anilhelvaci** [comment]
# Think
## Collateral Limit
* The limit for every pool will be stored in their corresponding `paramManager`
* The balance for every collateral type will be stored in `lendingPool`

### Challenge - How are we going to keep the `collateralBalance` updated?
We need a mechanism to update collateral balances as they change. Operations that effect collateral balances;
* New Borrows
* Adjust Balances
* Close Loan
* Liquidate Loan

`AdjustBalances` operation takes a special place as it can effect collateral balances in both ways. If the user wants to;
* `give: { Collateral: ... }`: This should increase the collateral balance
* `want: { Collateral: ... }`: This should decrease the collateral balance

### Solution - v1
Create a module called `balanceTracer.js` which contains code like below;
```js
export const makeBalanceTracer = () => {
  const balances = makeStore('Balances');

  const addNewBalanceType = brand => {
     balances.init(brand, AmountMath.makeEmpty(brand));
  };

  const updateBalance = (brand, amountToAdd, operation) => {
     const currentBalance = balances.get(brand);
     const newBalance = operation(currentBalance, amountToAdd);
     balances.set(brand, newBalance);
  };

  const getBalance = brand => balances.get(brand);

  return harden({
     addNewBalanceType,
     updateBalance,
     getBalance,
  });
};
```
And in `lendingPool.js`,
```js
const balanceTracer = makeBalanceTracer();

const addPoolType = (.....) => {
  ...
  const pm = makePoolManager(
  ....... ,
  balanceTracer,
  );
};
```

**Comment by anilhelvaci** [comment]
# Test Cases
Below are the test cases we tested so far for **Collateral Limits** as discussed above. Thinking about adding a test for a longer story.

## Collateral Limits

### Test Case - 1  
* Creator adds VAN and PAN pools
     * Total AgVAN collateral limit 10,000 units
* Alice borrows some PAN by putting 5,000 units of AgVAN
* Alice wants to adjust her loan by putting 6,000 units more AgVAN as collateral.
* Assert;
     * Should throw

### Test Case - 2 
* Creator adds VAN and PAN pools
     * Total AgVAN collateral limit 10,000 units
* Alice borrows some PAN by putting 5,000 units of AgVAN
* Bob wants to borrow some PAN by putting 7,500 units of AgVAN
* Assert;
     * Should throw

### Test Case - 3
* Creator adds VAN and PAN pools
     * Total AgVAN collateral limit 10,000 units
* Alice borrows some PAN by putting 5,000 units of AgVAN
* Alice adjusts her loan by putting in 2,500 units of AgVAN,
* Bob want to borrow by putting 5,000 units of AgVAN
* Assert;
     * Should throw

### Test Case - 4
* Creator adds VAN and PAN pools
     * Total AgVAN collateral limit 10,001 units
* Alice borrows PAN against 7,500 units of AgVAN
* Bob tries to borrow PAN against 5,000 units of AgVAN
* Assert;
     * Should throw
* Alice adjusts her loan by paying some PAN and receiving 2,500 AgVAN
* Bob tries to borrow
* Assert;
     * Should succeed

### Test Case - 5
* Creator adds VAN and PAN pools
     * Total AgVAN collateral limit 10,001 units
* Alice borrows PAN against 7,500 units of AgVAN
* Bob tries to borrow PAN against 5,000 units of AgVAN
* Assert;
     * Should throw
* Alice closes her loan
* Bob tries again
* Assert;
     * Should succeed

### Test Case - 6
* Creator adds VAN and PAN pools
     * Total AgVAN collateral limit 5,001 units
* Alice borrows PAN against 5,000 units of AgVAN
* Bob tries to borrow PAN against 5,000 units of AgVAN
* Assert;
     * Should throw
* Alice gets liquidated
* Bob tries again
* Assert;
     * Should succeed

### Test Case - 7
* Creator adds VAN and PAN pools
     * Total AgVAN collateral limit 5,000 units
* Alice tries to borrow PAN against 7,500 units of AgVAN
* Assert;
     * Should throw
* Creator increases the collateral limit to 7,501 units of AgVAN
* Alice tries again
* Assert;
     * Should succeed

## Borrowable
* Creator adds VAN and PAN pools
* Alice borrows PAN
* Creator marks PAN as non-borrowable
* Bob tries to borrow PAN
* Assert;
     * Should throw

## Usable As Collateral
* Creator adds VAN and PAN pools
* Alice borrows PAN using AgVAN as collateral
* Creator marks AgVAN as not usable as collateral
* Bob tries to borrow PAN using AgVAN
* Assert;
     * Should throw
